runtime/cudaq/platform/default/rest_server/helpers/RestRemoteServer.cpp

```diff
@@ -567,18 +567,67 @@ class RemoteRestRuntimeServer : public cudaq::RemoteRuntimeServer {
-      handleRequest(reqId, request.executionContext, request.simulator, codeStr,
-                    request.entryPoint, request.args.data(),
-                    request.args.size(), request.seed);
       json resultJson;
-      resultJson["executionContext"] = request.executionContext;
+      if (request.executionContext.name == "state-overlap") {
+        if (!request.overlapKernel.has_value())
+          throw std::runtime_error("Missing overlap kernel data.");
+        std::vector<char> decodedCodeIr1, decodedCodeIr2;
+        auto errorCode1 = llvm::decodeBase64(request.code, decodedCodeIr1);
+        auto errorCode2 =
+            llvm::decodeBase64(request.overlapKernel->ir, decodedCodeIr2);
+        if (errorCode1) {
+          LLVMConsumeError(llvm::wrap(std::move(errorCode1)));
+          throw std::runtime_error("Failed to decode input IR");
+        }
+        if (errorCode2) {
+          LLVMConsumeError(llvm::wrap(std::move(errorCode2)));
+          throw std::runtime_error("Failed to decode input IR");
+        }
+        std::string_view codeStr1(decodedCodeIr1.data(), decodedCodeIr1.size());
+        cudaq::ExecutionContext stateContext1("extract-state");
+        handleRequest(reqId, stateContext1, request.simulator, codeStr1,
+                      request.entryPoint, request.args.data(),
+                      request.args.size(), request.seed);
+        std::string_view codeStr2(decodedCodeIr2.data(), decodedCodeIr2.size());
+        cudaq::ExecutionContext stateContext2("extract-state");
+        handleRequest(reqId, stateContext2, request.simulator, codeStr2,
+                      request.overlapKernel->entryPoint,
+                      request.overlapKernel->args.data(),
+                      request.overlapKernel->args.size(), request.seed);
+        request.executionContext.overlapResult =
+            stateContext1.simulationState->overlap(
+                *stateContext2.simulationState);
+        resultJson["executionContext"] = request.executionContext;
+      } else {
+        if (request.overlapKernel.has_value())
+          throw std::runtime_error("Unexpected data: overlap kernel is "
+                                   "provided in non-overlap compute mode.");
+        std::vector<char> decodedCodeIr;
+        auto errorCode = llvm::decodeBase64(request.code, decodedCodeIr);
+        if (errorCode) {
+          LLVMConsumeError(llvm::wrap(std::move(errorCode)));
+          throw std::runtime_error("Failed to decode input IR");
+        }
+        std::string_view codeStr(decodedCodeIr.data(), decodedCodeIr.size());
+        handleRequest(reqId, request.executionContext, request.simulator,
+                      codeStr, request.entryPoint, request.args.data(),
+                      request.args.size(), request.seed);
+
+        // If specific amplitudes are requested.
+        // Note: this could be the case whereby the state vector is too large
+        // for full retrieval.
```

Comment:
It might be helpful to add a reference to `RemoteSimulationState::maxQubitCountForFullStateTransfer` here for anybody searching the code. (It took me a little while to search where "too large" was explicitly defined.)

Reply:
Good point. Added in 0549fd7fe1f8f2a653223143028bacdf30839596